### PR TITLE
Add dns() method and a callback for logging error before a retry

### PR DESCRIPTION
**Pull request by TyeMcQueen**

### CODE REVIEW

**Comment by ap** [comment]
Why limit the callback to `fixup` mode? Doesn’t it make just as much sense in `ping` mode?

**Comment by TyeMcQueen** [comment]
I no longer use DBI.  Closing ancient request to remove clutter.

**Comment by karenetheridge** [comment]
The fixup retry logging hook looks useful. But why is it called 'esub'?

**Review by theory — COMMENTED** [review]
LOL, apparently I reviewed this PR ages ago and forgot to submit it. Bunch of comments pending, now committed. Sorry for the delay. Do with them what you will.

**Comment by TyeMcQueen** [comment]
The error sub callback could be useful for `ping` mode.  It is less useful there as `ping` failing is usually just a stale connection rather than something more interesting that is more likely to be worth logging, hence probably why I didn't cover that case.

Feel free to come up with a different name for the error sub than "esub".  The other items had short names and that was the short name I came up with (and it isn't part of the public interface). (:

Ooh, more comments from years ago... I'd have to swap _fixup_run() back into my brain so for now I'll just say "maybe" it should also invoke the callback.

Just to be clear, since I'm no longer using any of this, I won't be making further changes.  But I'm happy to help with understanding the changes I did make, within reason. :)

**Comment by TyeMcQueen** [comment]
It looks to me like the code already includes a call to esub in _fixup_run().

File: lib/DBIx/Connector.pm [hunk 1]
```diff
@@ -854,6 +885,21 @@ or C<svp()>, the mode attribute will be set to the optional first parameter:
...
 In this way, you can reliably tell in what mode the code block is executing.
 
+For "fixup" mode, you may pass a second argument to C<mode()> that must be
+a CODE reference.  This routine will be called when a subsequent call to
+C<txn()> or C<run()> encounters an error that causes an operation to be
+retried.  Such a routine can be useful for logging transient errors that
+don't impact results but that might indicate infrastructure issues that
+could be investigated.
+
+  $conn->mode( fixup => sub {
```
[the remark this answers is not in the document]
Comment (ap) [reply lib/DBIx/Connector.pm:895]: I lean toward a separate accessor. It’s not clear to me why the `mode` mutator specifically would be the appropriate place to locate this functionality. Is that where someone looking for that kind of feature would go looking for it?

**Comment by karenetheridge** [comment]
Actually, do we need this callback at all? ..since I see DBI itself supports callbacks on any DBI method, including 'connected' -- e.g. see this example in its docs:

```
         my $dbh = DBI->connect($dsn, $username, $auth, {
             Callbacks => {
                 connected => sub {
                     shift->do(q{
                         SET SESSION sql_mode='ansi,strict_trans_tables,no_auto_value_on_zero';
                     });
                     return;
                 },
             }
         });
```

**Comment by TyeMcQueen** [comment]
As it so happens, I didn't want to log every time connected() was called.  I wanted to log whenever DBIx's fixup mode needed to retry.

But I stopped using DBI years ago so don't add it just for me. (:

File: lib/DBIx/Connector.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (ap) [reply lib/DBIx/Connector.pm:895]: More importantly, you can always run a block in `fixup` mode – regardless of the value of `$conn->{mode}`. So all of this strict validation is actually a mistake.

**Comment by ap** [comment]
@TyeMcQueen
> It is less useful there as `ping` failing is usually just a stale connection rather than something more interesting that is more likely to be worth logging, hence probably why I didn't cover that case.

More importantly, `fixup` mode is the only scenario where if the code block throws an exception, DBIx::Connector might not rethrow it. In all other cases, the caller is guaranteed to get to see the exception – but in `fixup` mode DBIx::Connector may swallow it.

So to answer to my question:

@ap
> Why limit the callback to `fixup` mode? Doesn’t it make just as much sense in `ping` mode?

No, in fact not.

@karenetheridge
> Actually, do we need this callback at all?

Yes we do. Because:

> I see DBI itself supports callbacks on any DBI method

That allows you to log the reconnection – but not to examine the exception that triggered it. That is only possible on the DBIx::Connector layer.

**Pull request closed without merging**